packages/firestore/test/unit/util/bundle.test.ts

```diff
@@ -0,0 +1,332 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { expect } from 'chai';
+import {
+  BundleReader,
+  SizedBundleElement,
+  toReadableStream
+} from '../../../src/util/bundle_reader';
+import { isNode } from '../../util/test_platform';
+
+function readableStreamFromString(
+  content: string,
+  bytesPerRead: number
+): ReadableStream {
+  return toReadableStream(new TextEncoder().encode(content), bytesPerRead);
+}
+
+function lengthPrefixedString(o: {}): string {
+  const str = JSON.stringify(o);
+  const l = new TextEncoder().encode(str).byteLength;
+  return `${l}${str}`;
+}
+
+// eslint-disable-next-line no-restricted-properties
+(isNode() ? describe.skip : describe)('readableStreamFromString()', () => {
+  it('returns stepping readable stream', async () => {
+    const encoder = new TextEncoder();
+    const s = readableStreamFromString('0123456789', 4);
+    const r = s.getReader();
+
+    let result = await r.read();
+    expect(result.value).to.deep.equal(encoder.encode('0123'));
+    expect(result.done).to.be.false;
+
+    result = await r.read();
+    expect(result.value).to.deep.equal(encoder.encode('4567'));
+    expect(result.done).to.be.false;
+
+    result = await r.read();
+    expect(result.value).to.deep.equal(encoder.encode('89'));
+    expect(result.done).to.be.false;
+
+    result = await r.read();
+    expect(result.value).to.be.undefined;
+    expect(result.done).to.be.true;
+  });
+});
+
+// eslint-disable-next-line no-restricted-properties
+(isNode() ? describe.skip : describe)('Bundle ', () => {
+  genericBundleReadingTests(1);
+  genericBundleReadingTests(4);
+  genericBundleReadingTests(64);
+  genericBundleReadingTests(1024);
+});
+
+function genericBundleReadingTests(bytesPerRead: number): void {
+  const encoder = new TextEncoder();
+  // Setting up test data.
+  const meta = {
+    metadata: {
+      id: 'test-bundle',
+      createTime: { seconds: 1577836805, nanos: 6 },
+      version: 1,
+      totalDocuments: 1,
+      totalBytes: 416
+    }
+  };
+  const metaString = lengthPrefixedString(meta);
+
+  const doc1Meta = {
```

Comment:
One benefit of not using JSON is that you could encore types here. It might help if you assigned a strict type here:

```
const doc1Meta : BundleMetadata = {}
```

Your test is currently mostly untyped. If you change the schema, your test will still compile and may even pass depending on if you update the test expectations. By enforcing types manually, you can help us to diagnose problems in the future.